linkerd/protocol/h2/src/main/scala/io/buoyant/linkerd/protocol/h2/H2RequestActiveTracer.scala

```diff
@@ -0,0 +1,147 @@
+ *
+ * A router that receives a tracer request with a max-forwards
+ * greater than 0 decrements the max-forwards header by 1 and
+ * forwards the tracer request to a downstream service that may perform
+ * additional processing of the request.
+ *
+ * @param endpoint the endpoint address of a downstream H2 service identified by a router
+ * @param namers namers used to evaluate a service name
+ * @param dtab base dtab used by the router
+ * @param routerLabel name of router
+ */
+class H2RequestActiveTracer(
+  endpoint: EndpointAddr,
+  namers: DstBindingFactory.Namer,
+  dtab: BaseDtab,
+  routerLabel: String
+) extends SimpleFilter[Request, Response] {
+
+  private[this] val MaxForwards = "max-forwards"
+  private[this] val AddRouterContextHeader = "l5d-add-context"
+
+  private[this] def getRequestTraceResponse(resp: Response, stopwatch: Stopwatch.Elapsed):Future[Response] = {
+    val routerCtxF = RouterContextBuilder(
+      routerLabel,
+      stopwatch,
+      DstPathCtx.current,
+      DstBoundCtx.current,
+      endpoint,
+      namers,
+      dtab)
+
+    routerCtxF.map { ctx =>
+      // format current router context
+      val content = s"""|
+                        |${ctx.formatRouterContext.trim}
+                        |""".stripMargin
+      if (resp.stream.isEmpty) {
+        // add router context data frame to empty stream
+        Response(resp.headers, Stream.const(Frame.Data(content, eos = true)))
+      } else {
+        val newStream = resp.stream.flatMap {
+          // read every frame of a stream
+          case trailerFrame: Trailers =>
+            // on trailers frame, prepend a data frame containing router context to trailer
+            val routerCtxFrame = Frame.Data(content, eos = false)
+            Seq(routerCtxFrame, trailerFrame)
+          case dataFrame: Data if dataFrame.isEnd =>
+            // on data frame with EOS flag to true, prepend new data frame with router context to
+            // EOS data frame
+            val newDataFrame = Frame.Data.copy(dataFrame, eos = false)
+            val routerCtxFrame = Data(content, eos = true)
+            Seq(newDataFrame, routerCtxFrame, Trailers())
+          case frame =>
+            // do nothing if we find a frame that doesn't indicate the end of a stream
+            Seq(frame)
+        }
+        Response(resp.headers, newStream)
+      }
+    }
+
+  }
+
+  override def apply(
+    request: Request,
+    service: Service[Request, Response]
+  ): Future[Response] = {
+    val maxForwards = Try(request.headers.get(MaxForwards).map(_.toInt))
+    val isAddRouterCtx = request.headers.get(AddRouterContextHeader) match {
+      case Some(v) => StringUtil.toBoolean(v)
+      case None => false
+    }
+
+    (request.method, maxForwards, isAddRouterCtx) match {
+      case (Method.Trace, Throw(_: NumberFormatException), _) =>
+        // MaxForwards is unparsable send a bad request response
+        Future.value(Response(Status.BadRequest, Stream.empty))
```

Comment:
Do we want to return an error string in the response, like we do in the h1 case?

Reply:
I was under the impression that the Status code would suffice, but I do see the point of being consistent. Also, it's useful information.